precompiles/ibc/ibc.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	pcommon "github.com/sei-protocol/sei-chain/precompiles/common"
+	"math/big"
+)
+
+const (
+	TransferMethod = "transfer"
+)
+
+const (
+	IBCAddress = "0x0000000000000000000000000000000000001008"
+)
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+func GetABI() abi.ABI {
+	abiBz, err := f.ReadFile("abi.json")
+	if err != nil {
+		panic(err)
+	}
+
+	newAbi, err := abi.JSON(bytes.NewReader(abiBz))
+	if err != nil {
+		panic(err)
+	}
+	return newAbi
+}
+
+type Precompile struct {
+	pcommon.Precompile
+	address        common.Address
+	transferKeeper pcommon.TransferKeeper
+	evmKeeper      pcommon.EVMKeeper
+
+	TransferID []byte
+}
+
+func NewPrecompile(transferKeeper pcommon.TransferKeeper, evmKeeper pcommon.EVMKeeper) (*Precompile, error) {
+	newAbi := GetABI()
+
+	p := &Precompile{
+		Precompile:     pcommon.Precompile{ABI: newAbi},
+		address:        common.HexToAddress(IBCAddress),
+		transferKeeper: transferKeeper,
+		evmKeeper:      evmKeeper,
+	}
+
+	for name, m := range newAbi.Methods {
+		switch name {
+		case TransferMethod:
+			p.TransferID = m.ID
+		}
+	}
+
+	return p, nil
+}
+
+// RequiredGas returns the required bare minimum gas to execute the precompile.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	methodID := input[:4]
+
+	method, err := p.ABI.MethodById(methodID)
+	if err != nil {
+		// This should never happen since this method is going to fail during Run
+		return 0
+	}
+
+	return p.Precompile.RequiredGas(input, p.IsTransaction(method.Name))
+}
+
+func (p Precompile) Run(evm *vm.EVM, caller common.Address, input []byte, value *big.Int) (bz []byte, err error) {
+	ctx, method, args, err := p.Prepare(evm, input)
+	if err != nil {
+		return nil, err
+	}
+
+	switch method.Name {
+	case TransferMethod:
+		return p.transfer(ctx, method, args)
+	}
+	return
+}
+func (p Precompile) transfer(ctx sdk.Context, method *abi.Method, args []interface{}) ([]byte, error) {
+	pcommon.AssertArgsLength(args, 6)
+
+	senderAddress, err := p.accAddressFromArg(ctx, args[0])
+	if err != nil {
+		return nil, err
+	}
+
+	receiverAddress, err := p.accAddressFromArg(ctx, args[1])
+	if err != nil {
+		return nil, err
+	}
+
+	port, ok := args[2].(string)
+	if !ok {
+		return nil, errors.New("port is not a string")
+	}
+
+	channelID, ok := args[3].(string)
+	if !ok {
+		return nil, errors.New("channelID is not a string")
+	}
+
+	denom := args[4].(string)
+	if denom == "" {
+		return nil, errors.New("invalid denom")
+	}
+
+	amount, ok := args[5].(*big.Int)
+	if !ok {
+		return nil, errors.New("amount is not a big.Int")
+	}
+
+	if amount.Cmp(big.NewInt(0)) == 0 {
+		// short circuit
+		return method.Outputs.Pack(true)
+	}
+
+	coin := sdk.Coin{
+		Denom:  denom,
+		Amount: sdk.NewIntFromBigInt(amount),
+	}
+
+	height := clienttypes.Height{}
```

Comment:
@codchen @Kbhat1 Not sure what values should be here...or should they be provided by the caller via ABI?

Reply:
will to defer to kartik here since it's IBC domain knowledge

Reply:
That's the timeout height, we can pass that in as a parameter in the transfer method and update the abi